GPflow/sgpr.py

```diff
@@ -0,0 +1,94 @@
+import tensorflow as tf
+import numpy as np
+from .model import GPModel
+from .param import Param
+from .densities import multivariate_normal
+from .mean_functions import Zero
+from . import likelihoods
+from .tf_hacks import eye
+
+class SGPR(GPModel):
+
+    def __init__(self, X, Y, kern, Z, mean_function=Zero()):
+        """
+        This Sparse Variational GP regression. The key reference is
+
+        @inproceedings{titsias2009variational,
+          title={Variational learning of inducing variables in sparse Gaussian processes},
+          author={Titsias, Michalis K},
+          booktitle={International Conference on Artificial Intelligence and Statistics},
+          pages={567--574},
+          year={2009}
+        }
+
+
+        X is a data matrix, size N x D
+        Y is a data matrix, size N x R
+        Z is a matrix of pseudo inputs, size M x D
+        kern, mean_function are appropriate GPflow objects
+
+
+        This method only works with a Gaussian likelihood.
+        """
+        likelihood = likelihoods.Gaussian()
+        GPModel.__init__(self, X, Y, kern, likelihood, mean_function)
+        self.Z = Param(Z)
+        self.num_data = X.shape[0]
+        self.num_latent = Y.shape[1]
+
+    def build_likelihood(self):
+        """
+        Constuct a tensorflow function to compute the bound on the marginal likelihood
+
+        """
+
+        num_inducing = tf.shape(self.Z)[0]
+        num_data = tf.shape(self.Y)[0]
+        output_dim = tf.shape(self.Y)[1]
+
+        err =  self.Y - self.mean_function(self.X)
+        Kdiag = self.kern.Kdiag(self.X)
+        Knm = self.kern.K(self.X, self.Z)
+        Kmm = self.kern.K(self.Z) + eye(num_inducing) * 1e-6
+        L = tf.cholesky(Kmm)
+        beta = 1./self.likelihood.variance
+
+        # Compute A
+        tmp = tf.matrix_triangular_solve(L, tf.transpose(Knm), lower=True)*tf.sqrt(1./self.likelihood.variance)
+        A_ = tf.matmul(tmp, tf.transpose(tmp))
+        A = A_ + eye(num_inducing)
+        LA = tf.cholesky(A)
+
+        tmp = tf.matrix_triangular_solve(L, tf.matmul(tf.transpose(Knm), err), lower=True)
+        b = tf.matrix_triangular_solve(LA, tmp, lower=True) * beta
+
+        #compute log marginal bound
+        bound = -0.5*tf.cast(num_data*output_dim, tf.float64)*np.log(2*np.pi)
```

Comment:
The computation of the bound could do with a comment explaining how the code relates to the equations in Titsias' paper.